xray/backends/pynio.py

```diff
@@ -0,0 +1,247 @@
+            # https://github.com/Unidata/netcdf4-python/pull/220
+            
+            data = np.asscalar(self.array.get_value())
+            #data = None
+        else:
+            data = self.array[key]
+            print data
+        return data
+
+
+def _nc4_values_and_dtype(var):
+    if var.dtype.kind == 'U':
+        # this entire clause should not be necessary with netCDF4>=1.0.9
+        if len(var) > 0:
+            var = var.astype('O')
+        dtype = str
+    elif var.dtype.kind in ['i', 'u', 'f', 'S']:
+        # use character arrays instead of unicode, because unicode suppot in
+        # netCDF4 is still rather buggy
+        data, dims = maybe_convert_to_char_array(var.data, var.dims)
+        var = Variable(dims, data, var.attrs, var.encoding)
+        dtype = var.dtype
+    else:
+        raise ValueError('cannot infer dtype for netCDF4 variable')
+    return var, dtype
+
+
+def _nc4_group(ds, group, mode):
+    if group in set([None, '', '/']):
+        # use the root group
+        return ds
+    else:
+        # make sure it's a string
+        if not isinstance(group, basestring):
+            raise ValueError('group must be a string or None')
+        # support path-like syntax
+        path = group.strip('/').split('/')
+        for key in path:
+            try:
+                ds = ds.groups[key]
+            except KeyError as e:
+                if mode != 'r':
+                    ds = ds.createGroup(key)
+                else:
+                    # wrap error to provide slightly more helpful message
+                    raise IOError('group not found: %s' % key, e)
+        return ds
+
+
+def _ensure_fill_value_valid(data, attributes):
+    # work around for netCDF4/scipy issue where _FillValue has the wrong type:
+    # https://github.com/Unidata/netcdf4-python/issues/271
+    if data.dtype.kind == 'S' and '_FillValue' in attributes:
+        attributes['_FillValue'] = np.string_(attributes['_FillValue'])
+
+
+def _force_native_endianness(var):
+    # possible values for byteorder are:
+    #     =    native
+    #     <    little-endian
+    #     >    big-endian
+    #     |    not applicable
+    # Below we check if the data type is not native or NA
+    if var.dtype.byteorder not in ['=', '|']:
+        # if endianness is specified explicitly, convert to the native type
+        data = var.data.astype(var.dtype.newbyteorder('='))
+        var = Variable(var.dims, data, var.attrs, var.encoding)
+        # if endian exists, remove it from the encoding.
+        var.encoding.pop('endian', None)
+    # check to see if encoding has a value for endian its 'native'
+    if not var.encoding.get('endian', 'native') is 'native':
+        raise NotImplementedError("Attempt to write non-native endian type, "
+                                  "this is not supported by the netCDF4 python "
+                                  "library.")
+    return var
```

Comment:
I think a bunch of these private functions are not needed. They seem to be NetCDF4-specific.